### DISCUSSION TITLE
Evaluation Guide for dplcompat storage backend

### DISCUSSION THREAD

**Discussion by arogge** [bareos/bareos#2063] [Show and tell]
# Evaluation Guide for [dplcompat storage backend](https://docs.bareos.org/TasksAndConcepts/StorageBackends.html#dplcompat-storage-backend)

## Prerequisites
This guide assumes that you know how to set up Bareos, know how to configure and use multiple storage devices and have successfully backed up to and restored from different storages.
Previous experience with the droplet backend might be helpful, but is not required.

To run the software, you will need a machine running a Linux or FreeBSD system supported by Bareos. Even though dplcompat works fine on Windows, this guide doesn't cover the Windows version.

To store data, you will need read-write access to an S3 bucket.

### Error reporting
Detailed error reporting is currently only available when debug logging is enabled. This is a limitation of the storage backend  API.

### Performance
As we have not yet done a lot of performance optimization, the `dplcompat` backend will currently provide slightly worse performance than the `droplet` backend.

### Installing packages
You will need at least the Bareos storage daemon version 24.0.0 or later, the package `bareos-storage-dplcompat` and s3cmd (either via the system package manager, pip or any other installation method).

> **NOTE**
> If you want to run other Bareos components (like the director) on the same machine, you can just install the required packages.

### configuring `s3cmd`
The dplcompat backend will call out to `s3cmd` to do the actual storage and retrieval of data. This will, of course, require a working setup of `s3cmd`. You can follow the [s3cmd-howto](https://s3tools.org/s3cmd-howto) to set up the program.

Once you have a working setup, copy `$HOME/.s3cfg` to `/etc/bareos/s3cfg` and ensure the Bareos user can read it (`chown bareos: /etc/bareos/s3cfg`).
You should now be able to successfully run `s3cmd` as the bareos user as follows:
```shell
# list buckets
sudo -u bareos s3cmd --config /etc/bareos/s3cfg ls

# get bucket information
sudo -u bareos s3cmd --config /etc/bareos/s3cfg info s3://backup
```

Initial performance tests have shown that the default chunk-size that `s3cmd` uses is way too small to provide good performance. To increase this to 250 MB (that was the sweet-spot we found during testing) change the `multipart-chunksize-mb` prameter in `/etc/bareos/s3cfg` to the value `250`.

### configuring the wrapper-script
The wrapper script can be configured using environment variables (these will be passed down from the storage daemon to the script).
Currently the following options are supported:

#### s3cmd_prog (autodetected by default)
Configure the full path to the `s3cmd` program to use.

#### s3cfg (no default)
Full path to the s3cfg file to read the configuration from.

#### bucket (default: "backup")
Name of the bucket to use.

#### base_url (default: "s3://${bucket}")
Override the base URL used to generate the object URLs from. When this is set, the bucket option is ignored.

#### storage_class (no default)
Defines the storage class of newly uploaded objects.

#### Testing the script
To test if the script works, you can run a few basic tests. As error reporting in Bareos itself is not yet providing a lot of information, this will allow you to address issues early, if neccessary.
As we need to inject the environment variables that Bareos will supply, we create another wrapper around the script that sets these.

```shell
#!/usr/bin/env bash
# set required variables...
s3cfg=/etc/bareos/s3cfg
bucket=your-bucket
# ... and defer to the real script
source /usr/lib/bareos/scripts/s3cmd-wrapper.sh
```
Make sure your script is readable/executable for the Bareos user.

Now you can use that for some basic testing
```shell
# connection test
sudo -u bareos ./test-wrapper.sh testconnection

# upload data
echo "Hello, World!" | sudo -u bareos ./test-wrapper.sh upload test 0000

# list objects (output: "0000 14")
sudo -u bareos ./test-wrapper.sh list test

# get object size (output: "14")
sudo -u bareos ./test-wrapper.sh stat test 0000

# download data (output: "Hello, World!")
sudo -u bareos ./test-wrapper.sh download test 0000

# remove an object
sudo -u bareos ./test-wrapper.sh remove test 0000
```

### configure the storage daemon
Create a device configuation in `/etc/bareos/bareos-sd.d/device/s3.conf` as follows:
```=
Device {
  Name = s3
  Media Type = s3
  Archive Device = S3 Object Storage
  Device Options = "iothreads=4"
                   ",ioslots=2"
                   ",chunksize=262144000"
                   ",program=s3cmd-wrapper.sh"
                   ",s3cfg=/etc/bareos/s3cfg"
                   ",bucket=your-bucket"
  Device Type = dplcompat
  LabelMedia = yes
  Random Access = yes
  AutomaticMount = yes
  RemovableMedia = no
  AlwaysOpen = no
  Maximum Concurrent Jobs = 1
}
```
The only changes that may be required here are on line 5. The Device Options are a comma-separated list of  key-value pairs. The example shown here is what provided best performance during our initial test:

#### iothreads
The number of concurrent threads that the backend will use for uploading. Setting this to 0 will disable threading (uploads will happen synchronously)

#### ioslots
The number of chunks per io-thread the backend will queue up before it blocks

#### chunksize
The maximum size of a chunk in bytes. These are the chunks that Bareos will pipe into `s3cmd`. While it makes sense to configure the `s3cmd` chunk size larger or equal to this, these are completely independent options.

#### program
The program that will be called for upload, download, etc. Either a full path starting with a `/` or relative to the storage daemons [Scripts Directory](https://download.bareos.org/experimental/PR-1790/BareosMainReference/Configuration/StorageDaemon.html#config-Sd_Storage_ScriptsDirectory) (on Debian this defaults to `/usr/lib/bareos/scripts`).

The settings `iothreads`, `ioslots` and `chunksize` will have an impact on the storage daemon's memory footprint. For each device configured, the peak memory usage will be $iothreads * ioslots * chunksize$ bytes.
As a result the sample above will require up to $2 * 4 * 262144000 = 2097152000$ bytes of memory (2000 MB).

> **WARNING**
> You must make sure you have enough memory available for this. Otherwise you machine may run into an out-of-memory condition.
> Especially when testing with director, storage daemon and PostgreSQL on the same machine, special care should be taken not to overcommit memory too much.

> **NOTE**
> Don't forget to restart the storage daemon after a configuration change.

### configure the director
Create a storage configuration in `/etc/bareos/bareos-dir.d/storage/s3.conf` as follows:
```=
Storage {
  Name = s3
  Address = 192.0.2.100
  Password = "AABBCCDDEEFFGGHHIIJJKKLLMMNN"
  Device = s3
  Media Type = s3
}
```
You need to adapt line 3 to match the address of your storage daemon (director and file daemons will use that address to connect) and line 4 to the storage password.

After a configuration reload, you can now status your new storage device to check that things are working by runnig `status storage=s3` in `bconsole`.

## Using the new storage
After setup is complete, you can now use your new storage just like every other storage: label volumes, run backup jobs to it, restore from there, etc.

**Comment by pstorz** [comment]
### Evaluating the dplcompat storage backend on Rocky Linux 9 on Hetzner Object Storage

This guide describes how to evaluate the dplcompat storage backend on Rocky Linux 9 using the Hetzner Object Storage offering as backup target.

### Setup your Hetzner Object Storage

* Setup your s3 keys for your object storage and follow [this guide](https://docs.hetzner.com/storage/object-storage/getting-started/generating-s3-keys
)

The result of the preparation is the following information:
* Access key
* Secret key
* bucket name


### Install s3cmd
```
dnf install -y epel-release
dnf install -y s3cmd
```
### Setup S3cmd

Use the following [this guide](https://docs.hetzner.com/storage/object-storage/getting-started/using-s3-api-tools/)

In my setup, the overview looks like this:
```
  Access Key: XXXX
  Secret Key: XXXX
  Default Region: US
  S3 Endpoint: hel1.your-objectstorage.com
  DNS-style bucket+hostname:port template for accessing a bucket: %(bucket)s.hel1.your-objectstorage.com
  Encryption password: XXXXXX
  Path to GPG program: /usr/bin/gpg
  Use HTTPS protocol: True
  HTTP Proxy server name: 
  HTTP Proxy server port: 0
```
Do not forget to copy the created s3cfg file to `/etc/bareos/s3cfg` and allow the user bareos to read it:

```
cp ~/.s3cfg /etc/bareos/s3cfg
chown bareos /etc/bareos/s3cfg
```

### Install the bareos dplcompat backend package:
These instructions are based on a basic installation of Bareos.

Install the bareos dplcompat backend package:

`dnf install -y bareos-storage-dplcompat`

### Test the s3cmd-wrapper.sh script

The dplcompat backend uses the `s3cmd-wrapper.sh` shell script to do the actual work on the s3 backend.
Therefore, the `s3cmd-wrapper.sh` script implements the following commands:

* testconnection
* list
* stat
* upload
* download
* remove

The configuration of the `s3cmd-wrapper.sh` script happens via environment variables.

To make sure that the s3cmd-wrapper.sh script will work with the configuration that we want to give, we have to run the wrapper script with the desired configuration set in the environment and then execute the script as user "bareos" like this:

`sudo s3cfg=/etc/bareos/s3cfg bucket=bucket-name -u bareos /usr/lib/bareos/scripts/s3cmd-wrapper.sh`

as typing this every time is quite cumersome, I created a bash alias that allows me to easily run the command as needed:

`alias runs3wrapper="sudo s3cfg=/etc/bareos/s3cfg bucket=bucket-name -u bareos /usr/lib/bareos/scripts/s3cmd-wrapper.sh"`

This allows to test every command of the wrapper script quite easy:

###testconnection
```
runs3wrapper testconnection
ERROR: Error parsing xml: Malformed error XML returned from remote server..  ErrorXML: b'<html><body><h1>501 Not Implemented</h1>\n.The server does not support the functionality required to fulfill the request.\n</body></html>\n'
ERROR: Error parsing xml: Malformed error XML returned from remote server..  ErrorXML: b'<html><body><h1>501 Not Implemented</h1>\n.The server does not support the functionality required to fulfill the request.\n</body></html>\n'
s3://bucket-name/ (bucket):
   Location:  hel1-prod1
   Payer:     none
   Ownership: none
   Versioning:none
   Expiration rule: none
   Block Public Access: none
   Policy:    none
   CORS:      none
   ACL:       p408981: FULL_CONTROL
```
Please note that the xml parser warnings have been no problem in my environment. Maybe some extension is missing on the tested object storage.

### upload
```

echo "Hello, World!" | runs3wrapper upload test 0000
upload: '<stdin>' -> 's3://bucket-name/test/0000' (0 bytes in 7.5 seconds, -1.00 B/s) [1 of 1]
```

### stat
```
runs3wrapper stat test 0000
14
```
### download
```
runs3wrapper download test 0000
Hello, World!
```
### remove
```
runs3wrapper remove test 0000
delete: 's3://bucket-name/test/0000'
```

### Configure the dplcompat backend in Bareos
If ever test above worked out as desired, we can finally configure the bareos dplcompat backend as follows:

* Create a **device** configuration file based on the installed example file:
```
cp -a /etc/bareos/bareos-sd.d/device/dplcompat.conf.example /etc/bareos/bareos-sd.d/device/dplcompat.conf
```
* Adapt the file and set the device options according to your setup. Set the "bucket"value as needed.
* Remove the "storage_class" setting for Hetzner.

 ```
 Device Options = "iothreads=4"
                   ",ioslots=2"
                   ",chunksize=262144000"
                   ",program=s3cmd-wrapper.sh"
                   ",s3cfg=/etc/bareos/s3cfg"
                   ",bucket=bucket-name"

```

* Create a **storage**configuration file based on the installed 
```cp -a /etc/bareos/bareos-dir.d/storage/dplcompat.conf.example /etc/bareos/bareos-dir.d/storage/dplcompat.conf
```
* Edit the newly created file  `/etc/bareos/bareos-dir.d/storage/dplcompat.conf` and adapt the values for **Adress** and **Password** as needed.
You can simply reuse the values from the preinstalled file `/etc/bareos/bareos-dir.d/storage/File.conf`

Finally restart the director and storage services:
```
systemctl restart bareos-dir bareos-sd
```
### Verify the dplcompat device

Start the bconsole and  check that the status command shows "Backend connection is working":
```
* status storage=dplcompat
[...]
Backend connection is working.
[...]
```

### Run the first backup
Make sure to set the storage for the test job:
```
* run job=backup-bareos-fd storage=dplcompat level=Full
* messages
 [...]
 2025-02-17 17:26:35 bareos-dir JobId 1: Created new Volume "Full-0001" in catalog.
 2025-02-17 17:26:35 bareos-sd JobId 1: Using Device "dplcompat" (S3) to write.
 2025-02-17 17:26:48 bareos-sd JobId 1: Labeled new Volume "Full-0001" on device "dplcompat" (S3).
 2025-02-17 17:26:48 bareos-sd JobId 1: Moving to end of data on volume "Full-0001"
 2025-02-17 17:26:49 bareos-sd JobId 1: Ready to append to end of Volume "Full-0001" size=214
 2025-02-17 17:26:49 bareos-sd JobId 1: Releasing device "dplcompat" (S3).
 2025-02-17 17:27:09 bareos-sd JobId 1: Elapsed time=00:00:34, Transfer rate=1.215 M Bytes/second
 2025-02-17 17:27:09 bareos-dir JobId 1: Insert of attributes batch table with 393 entries start
 2025-02-17 17:27:09 bareos-dir JobId 1: Insert of attributes batch table done
[...]
  Termination:            Backup OK
```

The predefined job successfully backed up 393 files to the dplcompat storage backend.
On the object storage, there was created a file `bucket-name/Full-0001/0000`.

### Run the first restore

* Start the  bconsole command and run the following command:
```
restore jobid=1 all done yes
```
This command will restore the whole backup created before and restore all below the directory `/tmp/bareos-restores`.


```
17-Feb 17:33 bareos-dir JobId 2: Version: 24.0.1~pre67.4ee24a825 (12 February 2025) Red Hat Enterprise Linux release 9.5 (Plow)
17-Feb 17:33 bareos-dir JobId 2: Start Restore Job RestoreFiles.2025-02-17_17.33.48_10
17-Feb 17:33 bareos-dir JobId 2: Connected Storage daemon at rocky9:9103, encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 bareos-dir JobId 2:  Encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 bareos-dir JobId 2: Using Device "dplcompat" to read.
17-Feb 17:33 bareos-dir JobId 2: Connected Client: bareos-fd at localhost:9102, encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 bareos-dir JobId 2:  Handshake: Immediate TLS 
17-Feb 17:33 bareos-dir JobId 2:  Encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 rocky9-fd JobId 2: Connected Storage daemon at rocky9:9103, encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 bareos-sd JobId 2: Version: 24.0.1~pre67.4ee24a825 (12 February 2025) Red Hat Enterprise Linux release 9.5 (Plow)
17-Feb 17:33 rocky9-fd JobId 2:  Encryption: TLS_CHACHA20_POLY1305_SHA256 TLSv1.3
17-Feb 17:33 bareos-sd JobId 2: Ready to read from volume "Full-0001" on device "dplcompat" (S3).
17-Feb 17:33 rocky9-fd JobId 2: Version: 24.0.1~pre67.4ee24a825 (12 February 2025) Red Hat Enterprise Linux release 9.5 (Plow)
17-Feb 17:33 bareos-sd JobId 2: Forward spacing Volume "Full-0001" to file:block 0:214.
17-Feb 17:33 bareos-sd JobId 2: End of Volume at file 0 on device "dplcompat" (S3), Volume "Full-0001"
17-Feb 17:33 bareos-sd JobId 2: End of all volumes.
17-Feb 17:33 bareos-sd JobId 2: Releasing device "dplcompat" (S3).
17-Feb 17:34 bareos-dir JobId 2: Bareos bareos-dir 24.0.1~pre67.4ee24a825 (12Feb25):
  Build OS:               Red Hat Enterprise Linux release 9.5 (Plow)
  JobId:                  2
  Job:                    RestoreFiles.2025-02-17_17.33.48_10
  Restore Client:         "bareos-fd" 24.0.1~pre67.4ee24a825 (12Feb25) Red Hat Enterprise Linux release 9.5 (Plow),redhat
  Start time:             17-Feb-2025 17:33:50
  End time:               17-Feb-2025 17:34:00
  Elapsed time:           10 secs
  Files Expected:         393
  Files Restored:         393
  Bytes Restored:         41,264,233
  Rate:                   4126.4 KB/s
  FD Errors:              0
  FD termination status:  OK
  SD termination status:  OK
  Bareos binary info:     Bareos community build (UNSUPPORTED): Get professional support from https://www.bareos.com
  Job triggered by:       User
  Termination:            Restore OK
```
Congratulations, the setup of the dplcompat backend was successful.

Now you can go on and configure the bareos system to fulfill your backup backup scheme.

**Comment by bruno-at-bareos** [comment]
## Warning if you are migrating an existing droplet device

You might have defined `chunksize=100M` in your droplet device.
For dplcompat you will have to translate that in Bytes.

**dplcompat** has to be defined like this: **`chunksize=104857600`**
